# Steam Gaming Dataset Exploration
By: Stefan Lauren

Date: Mar 2019

Background: Steam is one of the largest gaming networks in the world with over 100 million active gamers. The Steam dataset covers 109 million user accounts, 196 million friendships, 3 million groups, 384 million owned games, and a collective 1 million years of playtime.

The explanation of the data set could be found in https://steam.internet.byu.edu/

In [1]:
from pyspark import SparkConf, SparkContext, SQLContext
import pyspark.sql.functions as SqlF

In [2]:
try:
    sc = SparkContext()
    sqlContext = SQLContext(sc)
except ValueError:
    warnings.warn("SparkContext already exists in this scope")

## 1. Data Collection

In [3]:
# File location
file_player_summaries = "steam_gaming_small\Player_Summaries.csv"
file_games_publishers = "steam_gaming_small\Games_Publishers.csv"
file_games_genres = "steam_gaming_small\Games_Genres.csv"
file_games_developers = "steam_gaming_small\Games_Developers.csv"
file_games_1 = "steam_gaming_small\Games_1.csv"

In [4]:
# Load all csv
df_games_publishers = sqlContext.read.format("csv").option("inferSchema", 
                                                           True).option("header", True).load(file_games_publishers)

df_games_genres = sqlContext.read.format("csv").option("inferSchema", 
                                                           True).option("header", True).load(file_games_genres)

df_games_developers = sqlContext.read.format("csv").option("inferSchema", 
                                                           True).option("header", True).load(file_games_developers)

df_games_1 = sqlContext.read.format("csv").option("inferSchema", 
                                                           True).option("header", True).load(file_games_1)

df_player_summaries = sqlContext.read.format("csv").option("inferSchema", 
                                                           True).option("header", True).load(file_player_summaries)

## 2. Data Exploration

### 2.1 Game Publishers

In [5]:
df_games_publishers.printSchema()

root
 |-- appid: integer (nullable = true)
 |-- Publisher: string (nullable = true)



In [6]:
df_games_publishers.show(5)

+------+---------+
| appid|Publisher|
+------+---------+
|207990|     null|
|215220|     null|
|220824|     null|
|241600|     null|
|249310|     null|
+------+---------+
only showing top 5 rows



Game publishers table has appid as the id key for applications and publisher as the name of the publisher. There might be multiple rows with same appid, meaning that they publish an application together.

### 2.2 Games Genres

In [7]:
df_games_genres.printSchema()

root
 |-- appid: integer (nullable = true)
 |-- Genre: string (nullable = true)



In [8]:
df_games_genres.show(5)

+-----+-----+
|appid|Genre|
+-----+-----+
| 7290|  RPG|
| 8980|  RPG|
|18010|  RPG|
|18040|  RPG|
|24447|  RPG|
+-----+-----+
only showing top 5 rows



This is a table with the genre of application with id appid. An application may have more than one genre which are showed by multiple row with same appid.

### 2.3 Games Developers

In [9]:
df_games_developers.printSchema()

root
 |-- appid: integer (nullable = true)
 |-- Developer: string (nullable = true)



In [10]:
df_games_developers.show(5)

+------+---------+
| appid|Developer|
+------+---------+
|462530|       8i|
|452420|       M2|
|466530|       M2|
|366960|       UD|
|315060|      2SD|
+------+---------+
only showing top 5 rows



Developer is the name of the application's developer who made the application with id appid.

### 2.4 Games 1

In [11]:
df_games_1.printSchema()

root
 |-- steamid: long (nullable = true)
 |-- appid: integer (nullable = true)
 |-- playtime_2weeks: integer (nullable = true)
 |-- playtime_forever: integer (nullable = true)
 |-- dateretrieved: string (nullable = true)



In [12]:
df_games_1.show(5)

+-----------------+------+---------------+----------------+--------------------+
|          steamid| appid|playtime_2weeks|playtime_forever|       dateretrieved|
+-----------------+------+---------------+----------------+--------------------+
|76561198001291264|  8870|           null|            1392|2013-06-09 01:01:...|
|76561198001291264|   400|           null|             239|2013-06-09 01:01:...|
|76561198001291264|212910|           null|             130|2013-06-09 01:01:...|
|76561198001291264|   550|           null|           17547|2013-06-09 01:01:...|
|76561198001291264|   420|           null|             534|2013-06-09 01:01:...|
+-----------------+------+---------------+----------------+--------------------+
only showing top 5 rows



- steamid: user id
- appid: collection of applications that the user has
- playtime_2weeks: total amount of time during two weeks that the user spent on the application
- playtime_forever: total amount of time the user spent on the application
- dateretrieved: the first time the user got the application

### 2.5 Player Summaries

In [13]:
df_player_summaries.printSchema()

root
 |-- steamid: long (nullable = true)
 |-- personaname: string (nullable = true)
 |-- profileurl: string (nullable = true)
 |-- avatar: string (nullable = true)
 |-- avatarmedium: string (nullable = true)
 |-- avatarfull: string (nullable = true)
 |-- personastate: integer (nullable = true)
 |-- communityvisibilitystate: integer (nullable = true)
 |-- profilestate: integer (nullable = true)
 |-- lastlogoff: timestamp (nullable = true)
 |-- commentpermission: integer (nullable = true)
 |-- realname: string (nullable = true)
 |-- primaryclanid: long (nullable = true)
 |-- timecreated: timestamp (nullable = true)
 |-- gameid: integer (nullable = true)
 |-- gameserverip: string (nullable = true)
 |-- gameextrainfo: string (nullable = true)
 |-- cityid: string (nullable = true)
 |-- loccountrycode: string (nullable = true)
 |-- locstatecode: string (nullable = true)
 |-- loccityid: integer (nullable = true)
 |-- dateretrieved: timestamp (nullable = true)



This table has 23 features so I will not show the first rows as they are not visually easy to interpret. Instead, I will give entry example for every feature.

- steamid: user id (e.g. 76561198024350900)
- personaname: in-game alias name (e.g. DGJarski)
- profileurl: link to the profile (e.g. http://steamcommunity.com/profiles/76561198024350932/)
- avatar: profile picture url (e.g. http://media.steampowered.com/steamcommunity/public/images/avatars/00/0005392e08218406955be6df367225d4b7a74783.jpg)
- avatarmedium: medium size profile picture url (e.g. http://media.steampowered.com/steamcommunity/public/images/avatars/00/0005392e08218406955be6df367225d4b7a74783_medium.jpg)
- avatarfull: full size profile picture url (e.g. http://media.steampowered.com/steamcommunity/public/images/avatars/00/0005392e08218406955be6df367225d4b7a74783_full.jpg)
- personastate: - (e.g. 0)
- communityvisibilitystate: - (e.g. 1)
- profilestate: - (e.g. 1)
- lastlogoff: the last time the user log off (e.g. 12/7/2011  2:31:52 AM)
- commentpermission: - (e.g. 2)
- realname: user real name (e.g. Brandon Taylor)
- primaryclanid: group id (e.g. 103582791429650000)
- timecreated: the time this account is created (e.g. 12/29/2004  9:00:06 AM)
- gameid: id of the game the user was playing by the time the data were collected (e.g. 207890)
- gameserverip: ip the user was used by the time the data were collected, by default it is 0.0.0.0:0 (e.g. 188.127.246.124:27705)
- gameextrainfo: - (e.g. Football Manager 2013)
- cityid: - (e.g. "")
- loccountrycode: country code (e.g. US)
- locstatecode: residence state (e.g. MI)
- loccityid: city (e.g. 2082)
- dateretrieved: timestamp (e.g. 3/8/2013  1:48:13 AM)

### 2.6 Count the number of games per `publisher` and per `genre`

To get the number of games per publisher, Games_Publishers table will be used. To get the number of games per genre, Games_Genre table will be used. The operations that I need are grouping the publisher and genre column; and count the number of application (appid) for every group.

In [14]:
# Number of games per publisher
df_games_publishers.createOrReplaceTempView("publishers")
games_per_publisher = sqlContext.sql("""
    SELECT Publisher, COUNT(appid) as Publisher_Count
    FROM publishers
    GROUP BY Publisher
    ORDER BY Publisher_Count desc
""")

games_per_publisher.show(games_per_publisher.count())

+--------------------+---------------+
|           Publisher|Publisher_Count|
+--------------------+---------------+
|                null|           2627|
|             Ubisoft|            384|
|                SEGA|            349|
|Dovetail Games - ...|            279|
| Paradox Interactive|            246|
|  Disney Interactive|            226|
|          Activision|            221|
|Feral Interactive...|            221|
|              Degica|            190|
|        Nordic Games|            164|
|         Square Enix|            153|
|            KISS ltd|            145|
|Feral Interactive...|            143|
|Wizards of the Co...|            135|
|      Strategy First|            134|
|            2K Games|            120|
|              Capcom|            115|
|Warner Bros. Inte...|            108|
|Kalypso Media Dig...|            104|
|         Deep Silver|             98|
|       Sekai Project|             95|
|  Night Dive Studios|             94|
|      Dovetail Games|   

#### Observation
The highest count is given by the null value, which I assume represents the indie publishers.

In [66]:
# Number of games per genre
df_games_genres.createOrReplaceTempView("genres")
games_per_genre = sqlContext.sql("""
    SELECT Genre, COUNT(appid) as Genre_Count
    FROM genres
    GROUP BY Genre
    ORDER BY Genre_Count desc
""")

games_per_genre.show(games_per_genre.count())

+--------------------+-----------+
|               Genre|Genre_Count|
+--------------------+-----------+
|               Indie|       7982|
|              Action|       7126|
|           Adventure|       4517|
|            Strategy|       3953|
|              Casual|       3939|
|          Simulation|       3870|
|                 RPG|       3147|
|        Free to Play|       1172|
|        Early Access|        901|
|Massively Multipl...|        748|
|              Racing|        619|
|              Sports|        604|
|Design & Illustra...|        240|
|           Utilities|        207|
|      Web Publishing|        146|
|Animation & Modeling|        112|
|    Audio Production|        112|
|   Software Training|         82|
|           Education|         73|
|    Video Production|         69|
|       Photo Editing|         46|
|          Accounting|          4|
+--------------------+-----------+



#### Observation

As I suspected, Indie games has the highest count followed by Action games with small difference.

### 2.7 Find day and hour when most new accounts were created

It will be based only on Player_Summaries table. I will find the highest count of newly created accounts based on the day only, the hour only, and the datetime.

In [15]:
# Remove null values
df_player_summaries = df_player_summaries.na.drop(subset=["timecreated"])
df_player_summaries.createOrReplaceTempView("player_summaries")

In [16]:
# Find a day when most new accounts were created
sqlContext.sql("""
  SELECT to_date(timecreated) as Start_Date, 
         to_date(date_add(timecreated,1)) as End_Date, 
         COUNT(steamid) as Count_New_Accounts
  FROM player_summaries
  WHERE timecreated >= to_date(timecreated) and timecreated < to_date(date_add(timecreated,1))
  GROUP BY Start_Date
  ORDER BY Count_New_Accounts desc
  LIMIT 5
""").show()

+----------+----------+------------------+
|Start_Date|  End_Date|Count_New_Accounts|
+----------+----------+------------------+
|2012-12-25|2012-12-26|              3682|
|2012-11-28|2012-11-29|              3366|
|2012-12-26|2012-12-27|              2691|
|2012-12-24|2012-12-25|              2634|
|2012-12-22|2012-12-23|              2508|
+----------+----------+------------------+



The day when most new accounts were created is between 2012/12/25 and 2012/12/26 with 3682 accounts newly created. Interestingly, five top results are all from year 2012 and between the end of November to the end of December on that year.

In [17]:
# Find the hour when most new accounts were created
sqlContext.sql("""
  SELECT hour(timecreated) as Start_Hour, 
         (hour(timecreated)+1) as End_Hour, 
         COUNT(steamid) as Count_New_Accounts
  FROM player_summaries
  GROUP BY Start_Hour
  ORDER BY Count_New_Accounts desc
  LIMIT 5
""").show()

+----------+--------+------------------+
|Start_Hour|End_Hour|Count_New_Accounts|
+----------+--------+------------------+
|        10|      11|             14950|
|        11|      12|             14660|
|         9|      10|             14368|
|         8|       9|             14167|
|        12|      13|             13975|
+----------+--------+------------------+



From the query, we could see that between 10AM to 11AM is the period when there were the highest number of newly created account. Moreover, we could say that most people created their account in the morning between 8AM to 13PM.

Finally, to find the datetime when most new accounts were created, I will extend the player summaries dataframe to include rounded hour 'timecreated' column as 'start_timecreated' and rounded hour plus 1 'timecreated' column as 'end_timecreated'.

In [18]:
df_player_summaries = df_player_summaries.withColumn("start_timecreated", 
    SqlF.from_unixtime(SqlF.unix_timestamp("timecreated","yyyy-MM-dd'T'HH:mm:ss"),"yyyy-MM-dd HH:00:00"))

df_player_summaries = df_player_summaries.withColumn("end_timecreated", 
    SqlF.from_unixtime(SqlF.unix_timestamp("timecreated","yyyy-MM-dd'T'HH:mm:ss") + 3600,"yyyy-MM-dd HH:00:00"))

df_player_summaries.createOrReplaceTempView("extended_player_summaries")

In [19]:
# Find the datetime when most new accounts were created
sqlContext.sql("""
  SELECT start_timecreated as Start_Time, 
         end_timecreated as End_Time, 
         COUNT(steamid) as Count_New_Accounts
  FROM extended_player_summaries
  WHERE timecreated >= start_timecreated and timecreated < end_timecreated
  GROUP BY Start_Time, End_Time
  ORDER BY Count_New_Accounts desc
  Limit 5
""").show()

+-------------------+-------------------+------------------+
|         Start_Time|           End_Time|Count_New_Accounts|
+-------------------+-------------------+------------------+
|2012-12-25 10:00:00|2012-12-25 11:00:00|               292|
|2012-12-25 08:00:00|2012-12-25 09:00:00|               279|
|2012-12-25 11:00:00|2012-12-25 12:00:00|               230|
|2012-11-28 11:00:00|2012-11-28 12:00:00|               219|
|2012-11-28 10:00:00|2012-11-28 11:00:00|               216|
+-------------------+-------------------+------------------+



It was on Christmas 2012 when there were a lot of newly created accounts in the morning, especially between 10AM and 11AM.

#### Observation

There might be special events that occurred between the end of November 2012 to the end of December 2012 with the peak of the event was on the Christmas Day (December 25th, 2012). It is concluded this way as the density of newly created accounts is considerably high on that period.

## 3. Rating Prediction Model

In this section, I will try to build a model to predict the rating of a game.

In [20]:
from pyspark.ml.feature import OneHotEncoder, StringIndexer, VectorAssembler
from pyspark.ml.regression import LinearRegression, DecisionTreeRegressor
from pyspark.ml.evaluation import RegressionEvaluator

In [21]:
# Read App_ID_Info.csv
file_app = "steam_gaming_small\App_ID_Info.csv"
df_app = sqlContext.read.format("csv").option("inferSchema", 
                                                           True).option("header", True).load(file_app)

The rating of a game is affected by some factors outside of the gameplay. Some of those factors are the price, age requirement, multiplayer feature, genre, publishers and developer. Therefore, we need to combine several table to get the field that we need.

In [22]:
df_games_publishers.createOrReplaceTempView("publishers")
df_games_genres.createOrReplaceTempView("genres")
df_games_developers.createOrReplaceTempView("developers")
df_app.createOrReplaceTempView("application")

In [23]:
df_application = sqlContext.sql("""
  SELECT application.Price, application.Required_Age, application.Is_multiplayer
  , genres.Genre, publishers.Publisher, developers.Developer, application.Rating
  FROM application LEFT JOIN genres
  ON application.appid == genres.appid
  LEFT JOIN publishers
  ON application.appid == publishers.appid
  LEFT JOIN developers
  ON application.appid == developers.appid
  WHERE application.Rating != -1
""")

# Drop NULL values
df_application = df_application.na.drop()

Along with removing rows with NULL values, observations with rating equals to -1 are removed as well as those observations would hinder the training of the model. (-1 is equal to not rated yet)

In [24]:
df_application.printSchema()

root
 |-- Price: string (nullable = true)
 |-- Required_Age: integer (nullable = true)
 |-- Is_multiplayer: integer (nullable = true)
 |-- Genre: string (nullable = true)
 |-- Publisher: string (nullable = true)
 |-- Developer: string (nullable = true)
 |-- Rating: string (nullable = true)



We can see that Price and Rating have type of string, but we want them to be numerical. Therefore, below is a step to change the column type.

In [25]:
df_application = df_application.withColumn('Price', df_application.Price.cast('float'))
df_application = df_application.withColumn('Rating', df_application.Rating.cast('integer'))

In [26]:
df_application.printSchema()

root
 |-- Price: float (nullable = true)
 |-- Required_Age: integer (nullable = true)
 |-- Is_multiplayer: integer (nullable = true)
 |-- Genre: string (nullable = true)
 |-- Publisher: string (nullable = true)
 |-- Developer: string (nullable = true)
 |-- Rating: integer (nullable = true)



In [27]:
df_application.show(5)

+-----+------------+--------------+--------------------+----------------+--------------------+------+
|Price|Required_Age|Is_multiplayer|               Genre|       Publisher|           Developer|Rating|
+-----+------------+--------------+--------------------+----------------+--------------------+------+
|  0.0|           0|             1|Massively Multipl...|         Ubisoft|Ivory Tower in co...|    71|
|  0.0|           0|             1|              Racing|         Ubisoft|Ivory Tower in co...|    71|
|  0.0|           0|             1|              Action|         Ubisoft|Ivory Tower in co...|    71|
|  0.0|           0|             0|           Adventure|Frictional Games|    Frictional Games|    67|
|  0.0|           0|             0|               Indie|Frictional Games|    Frictional Games|    67|
+-----+------------+--------------+--------------------+----------------+--------------------+------+
only showing top 5 rows



Now we have the DataFrame to predict the rating. However, there are three categorical features: Genre, Publisher and Developer that we need to preprocess into one hot representation before modelling.

In [28]:
def processCategorical(dataFrame, inputColIndex, outputColIndex, outputColOneHot):
    stringIndexer = StringIndexer(inputCol=inputColIndex, outputCol=outputColIndex)
    model = stringIndexer.fit(dataFrame)
    indexed = model.transform(dataFrame)
    encoder = OneHotEncoder(dropLast=False, inputCol=outputColIndex, outputCol=outputColOneHot)
    encoded = encoder.transform(indexed)
    
    return encoded

In [29]:
df_application_transformed = processCategorical(df_application, 'Genre', 'Genre_index', 'Genre_vec')
df_application_transformed = processCategorical(df_application_transformed, 'Publisher', 'Publisher_index', 'Publisher_vec')
df_application_transformed = processCategorical(df_application_transformed, 'Developer', 'Developer_index', 'Developer_vec')

In [30]:
df_application_transformed.printSchema()

root
 |-- Price: float (nullable = true)
 |-- Required_Age: integer (nullable = true)
 |-- Is_multiplayer: integer (nullable = true)
 |-- Genre: string (nullable = true)
 |-- Publisher: string (nullable = true)
 |-- Developer: string (nullable = true)
 |-- Rating: integer (nullable = true)
 |-- Genre_index: double (nullable = false)
 |-- Genre_vec: vector (nullable = true)
 |-- Publisher_index: double (nullable = false)
 |-- Publisher_vec: vector (nullable = true)
 |-- Developer_index: double (nullable = false)
 |-- Developer_vec: vector (nullable = true)



Now, we have the data we need to start modelling. As we want to predict rating, this will be a regression problem. First, let us try the linear regression method.

### 3.1 Linear Regression

In [31]:
# Prepare the Features and Label
vectorAssembler = VectorAssembler(inputCols = ['Price', 'Required_Age', 'Is_multiplayer', 
                                               'Genre_index', 'Publisher_index', 'Developer_index'], 
                                  outputCol = 'Features')
df_application_transformed_vec = vectorAssembler.transform(df_application_transformed)
df_application_transformed_vec = df_application_transformed_vec.select(['Features', 'Rating'])
df_application_transformed_vec.show(5)

+--------------------+------+
|            Features|Rating|
+--------------------+------+
|[0.0,0.0,1.0,10.0...|    71|
|[0.0,0.0,1.0,7.0,...|    71|
|(6,[2,5],[1.0,201...|    71|
|[0.0,0.0,0.0,3.0,...|    67|
|[0.0,0.0,0.0,1.0,...|    67|
+--------------------+------+
only showing top 5 rows



In [32]:
# Split into training set and testing set
splits = df_application_transformed_vec.randomSplit([0.7, 0.3])
train_df = splits[0]
test_df = splits[1]

In [33]:
lr = LinearRegression(featuresCol = 'Features', labelCol='Rating', maxIter=10, regParam=0.1, elasticNetParam=0.1)
lr_model = lr.fit(train_df)

In [34]:
trainingSummary = lr_model.summary

print("Train RMSE: %f" % trainingSummary.rootMeanSquaredError)
print("Train R2: %f" % trainingSummary.r2)

Train RMSE: 10.608134
Train R2: 0.063361


The linear regression model could not perform well. The R2 metric is very low, meaning that the model cannot explain the Rating.

In [35]:
lr_predictions = lr_model.transform(test_df)
lr_predictions.show(10)

+--------------------+------+-----------------+
|            Features|Rating|       prediction|
+--------------------+------+-----------------+
|(6,[0,4],[19.9899...|    91|74.63756906023482|
|(6,[0,4],[19.9899...|    91|74.63756906023482|
|(6,[0,4],[19.9899...|    91|74.62878065969153|
|(6,[0,4],[39.9900...|    72|76.60976954904056|
|(6,[0,5],[9.98999...|    82| 73.6154480174183|
|(6,[0,5],[9.98999...|    91|73.24113848103666|
|(6,[0,5],[9.98999...|    72|69.39345662911357|
|(6,[0,5],[9.98999...|    58|69.01914709273193|
|(6,[0,5],[19.9899...|    86|73.39605127855022|
|(6,[2,5],[1.0,450...|    85|71.23140581998685|
+--------------------+------+-----------------+
only showing top 10 rows



In [36]:
lr_evaluator = RegressionEvaluator(predictionCol="prediction", labelCol="Rating",metricName="r2")
test_result = lr_model.evaluate(test_df)

print("Test RMSE: %g" % test_result.rootMeanSquaredError)
print("Test R2: %g" % lr_evaluator.evaluate(lr_predictions))

Test RMSE: 10.6734
Test R2: 0.0771897


The performance on the test set is not good as well. Let's try another regression method to compare the result.

### 3.2 Decision Tree Regression

In [37]:
# Prepare the Features and Label
vectorAssembler = VectorAssembler(inputCols = ['Price', 'Required_Age', 'Is_multiplayer', 
                                               'Genre_vec', 'Publisher_vec', 'Developer_vec'], 
                                  outputCol = 'Features')
df_application_transformed_vec = vectorAssembler.transform(df_application_transformed)
df_application_transformed_vec = df_application_transformed_vec.select(['Features', 'Rating'])
df_application_transformed_vec.show(5)

+--------------------+------+
|            Features|Rating|
+--------------------+------+
|(2141,[2,13,15,10...|    71|
|(2141,[2,10,15,10...|    71|
|(2141,[2,3,15,101...|    71|
|(2141,[6,76,860],...|    67|
|(2141,[4,76,860],...|    67|
+--------------------+------+
only showing top 5 rows



In [38]:
# Split into training set and testing set
splits = df_application_transformed_vec.randomSplit([0.7, 0.3])
train_df = splits[0]
test_df = splits[1]

In [39]:
dt = DecisionTreeRegressor(featuresCol ='Features', labelCol = 'Rating')
dt_model = dt.fit(train_df)

In [40]:
dt_predictions = dt_model.transform(test_df)
dt_predictions.show(10)

+--------------------+------+-----------------+
|            Features|Rating|       prediction|
+--------------------+------+-----------------+
|(2141,[0,1,2,3,15...|    82|71.67455197132617|
|(2141,[0,1,2,3,15...|    59|71.67455197132617|
|(2141,[0,1,2,3,15...|    88|79.66981132075472|
|(2141,[0,1,2,3,15...|    80|79.66981132075472|
|(2141,[0,1,2,3,16...|    81|71.67455197132617|
|(2141,[0,1,2,3,16...|    65|71.67455197132617|
|(2141,[0,1,2,3,17...|    88|          87.4375|
|(2141,[0,1,2,3,17...|    88|          87.4375|
|(2141,[0,1,2,3,21...|    86|79.66981132075472|
|(2141,[0,1,2,3,22...|    54|79.66981132075472|
+--------------------+------+-----------------+
only showing top 10 rows



In [41]:
dt_evaluator = RegressionEvaluator(predictionCol="prediction", labelCol="Rating", metricName="rmse")
rmse = dt_evaluator.evaluate(dt_predictions)
print("Test RMSE = %g" % rmse)

dt_evaluator = RegressionEvaluator(predictionCol="prediction", labelCol="Rating", metricName="r2")
print("Test R2: %g" % dt_evaluator.evaluate(dt_predictions))

Test RMSE = 10.6168
Test R2: 0.0754085


Using Decision Tree Regression, the RMSE and R squared are almost the same as Linear Regression method.

### 3.3 Conclusion and Future Works

During my experiments, both regression method could not perform well. Some reasons behind it would be:
- Lack of data
- Incorrect feature representation
- Unsuitable regression methods

For future works, several things that might improve the results are:
- Data augmentation
- Feature engineering